### DISCUSSION TITLE
Site Reference Metadata interface

### DISCUSSION THREAD

**Discussion by MikeNeilson** [opendcs/opendcs#644] [Ideas]
So CWMS contains a rather useful feature, Location Levels, that is unfortunately named as it's more than just levels. But anyways, the concept is for data that is generally static but *may* vary in time. OpenDCS has a concept of "Seasonal" values but the Location Levels, that I will now refer to as Site Reference Metadata, or just RefData so save my fingers, go beyond that.

 The primary example, and likely where the name originally came from would is pertinent data about a dam.
- Bottom of inlet
- Spillway crest
- Top of Dam
- Gross Pool
- Top of Conservation (this one gets tricky)
- etc

Bottom of inlet may never change, but could be useful in a computation to generically access.

Grosspool does change, but only when new bathemetry is taken, or some random water district decides to argue for making the spillway crest higher. But for say a "% full" pool calculation being able to lookup the Gross pool by the Site name instead of an arbitrary property allows for a group.

Top Con gets complicated in some USACE districts, like SPK where it is calculated daily (or in the case of Folsom sometimes 4x a day) and backed by an actual time series.

For others Top Con is a simple sequence within the window of the water year and may not change often but could with changes to the project. 

USACE districts have also used the interface to record various statistics such as peaks, mins, maxes, that get used in calculations or reports.

Thus the reference data is information the varies over time, but it not in itself a time series, but a full time series matching some need can be made from it, if necessary.

Each piece of RefData comes with an effective date and some information about exactly *what* it is. Additionally since the general idea is that the data is used for calculations with time series one can retrieve a "time series" of appropriate values. You can expand the data to the desired window and interval as needed. Or just pull the current value, like say your creating a "Dam Chart" and want reference markers for Top of Dam, Spillway, inlets, etc.

I would like to start implementing this concept into OpenDCS. Talking to the USACE people during one of our meetings we came up with "Site Reference Metadata" as seen in the title. Not totally set on it but currently seems the most appropriate.

Additionally while it *may* work to just take the USACE concept, I can pull more docs later for reference, it think it would be best to review that, see what HDB has (if anything similar) and discuss any perceived or actual short comings with the USACE design and make improvements.

If we can design a solid interface, we should be able to get any implementation to match the required data output fairly easily.

Also, no guarantees, but I can probably talk HEC it to donating the code to OpenDCS to take avoid both odd dependency cycles and duplicate work, but that will depend on the exact design with come up with.

HDB  and USACE maintainers have discussed merging our database schemas. Moving our required concepts into the OpenDCS-Postgres implementation and taking all of our lessons learned in naming and design is, I think, the best way to make that even possible and at a minimum gives all OpenDCS users a superior data processing experience which is what the project should strive to do anyways.

@opendcs/opendcs-core-devs @opendcs/opendcs-triage @perrymanmd @zamusacelre @oskarhurst @msweier 
Feel free to tag anyone in a comment if you think they will add value to this converstation.

**Comment by agilmore2** [comment]
Interesting. HDB has several site reference tables like this, originally
developed for storing consumptive use and natural flow data, including
agricultural statistics.

Not sure we've shared an ERD before, and there's a couple of tables on here
that are obsolete, but take a look at the REF_SITE_ATTR table and
associated tables. It's pretty flexible.

**Reply by MikeNeilson** [reply]
Good to know, unfortunately it appears the the USBR github got murdered *while* it's getting moved instead of after, and the hdb repo isn't on the new github org yet. Too bad I hadn't forked it.

**Reply by MikeNeilson** [reply]
https://github.com/MikeNeilson/HDB/blob/b49137a3b9ff72414eb5b042c984d2913cf30eed/oracle_script/SCHEMA/BASE_SCRIPTS/tbl.ddl#L1439

**Comment by MikeNeilson** [comment]
@oskarhurst  FYI, some additional "reference metadata" information to guide future design choices.
